### PR TITLE
Refactor wheel tests into individual tests

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**
Refactor how we run wheel tests so that each test is a separate entity, rather than just having all tests crammed into test-wheel.sh. In particular, refactor Linux testing to create a temporary Docker image that has the wheel installed, so that we will (on Linux) be able to create tests that need to manipulate the test environment without such manipulation affecting other tests.

- - -
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)

### CODE REVIEW

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please

Okay, I think the kinks have been ironed out. Seems to be working locally, at least.

**Comment by mwoehlke-kitware** [comment]
+@jwnimmer-tri for feature review, please.

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

...just so there's no question that it works. It *should*, it succeeded previously and the Linux code paths have not (AFAIK) been touched since then, but let's make sure, and also make it easier to verify that it's working.

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 6 of 8 files at r1, 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @mwoehlke-kitware)

___
*[`tools/wheel/macos/test-wheel.sh` line 35 at r3](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEGYr3W0uNYx9cL58dB:-NEGYr3W0uNYx9cL58dC:bolcrr6) ([raw file](https://github.com/robotlocomotion/drake/blob/974dd7e7abf7c37d04540c23ff3ff25405598d36/tools/wheel/macos/test-wheel.sh#L35)):*
> ```Shell
> 
> for t in "$@"; do
>     "$resource_root/test/test-wheel.sh" "$resource_root/test/$t" "$wheel"
> ```

It's too confusing to have multiple programs named "test-wheel" that are not equivalent.

This macos `test-wheel` runs _all_ the tests, but this called-into `test-wheel` runs just one test.

We need to use different filenames for the two different layers of responsibility, to make it easier to understand the software layering.
___
*[`tools/wheel/wheel_builder/common.py` line 23 at r3](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEGX5Au7sEs6xtF208n:-NEGX5Au7sEs6xtF208o:bh0nmzd) ([raw file](https://github.com/robotlocomotion/drake/blob/974dd7e7abf7c37d04540c23ff3ff25405598d36/tools/wheel/wheel_builder/common.py#L23)):*
> ```Python
>     'tests/size-test.py',
>     'tests/pyi-install-test.py',
> ]
> ```

As we have it here, the platform driver programs (common, linux, macos) list their tests one by one and then plumb that list down through a few branchy/loopy layers of bash scripts.

Having tests committed to git that accidentally don't actually get run in CI is one of my nightmares.  It's difficult to notice when it happens, and can lead to us tagging broken code as stable.

The lists here are a tiny bit brittle in that perhaps someone would add a new test program down the road, but fail to plumb it through correctly.  Hopefully local testing and code review would be enough to notice.

More importantly, though, there is just a lot of boilerplate to document, understand, and keep up to date.  It took me a while to follow the argument-passing conventions and control flow though the code, especially the bash code.

Here's a counter proposal:

(1) The drivers (common, linux, macos) do not hard-code a list of tests.

(2) Instead, all test programs are always run.  The infrastructure code will glob (and sort) files in the `tests` folder and run all of those.

(3) Test that are not relevant (e.g., exercising `apt` commands while on `macOS`) detect as much and embed a trivially-pass short-circuit within the test itself.

In short, we _encapsulate_ the "does this test case apply?" logic within the test itself, instead of the top-level driver.

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

**Review by mwoehlke-kitware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @jwnimmer-tri)

___
*[`tools/wheel/macos/test-wheel.sh` line 35 at r3](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEGYr3W0uNYx9cL58dB:-NEHoM98B6Jz0p2s8rz_:b-j3zyph) ([raw file](https://github.com/robotlocomotion/drake/blob/974dd7e7abf7c37d04540c23ff3ff25405598d36/tools/wheel/macos/test-wheel.sh#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It's too confusing to have multiple programs named "test-wheel" that are not equivalent.

This macos `test-wheel` runs _all_ the tests, but this called-into `test-wheel` runs just one test.

We need to use different filenames for the two different layers of responsibility, to make it easier to understand the software layering.
</blockquote></details>

Done. (The old `macos/test-wheel.sh` was basically just a bash version of `_test_wheel` from `linux.py`. The logic has been moved to `_test_wheel` in `macos.py`.)
___
*[`tools/wheel/wheel_builder/common.py` line 23 at r3](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEGX5Au7sEs6xtF208n:-NEHoeD586JhnODxXqu1:b-iyu9rz) ([raw file](https://github.com/robotlocomotion/drake/blob/974dd7e7abf7c37d04540c23ff3ff25405598d36/tools/wheel/wheel_builder/common.py#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

As we have it here, the platform driver programs (common, linux, macos) list their tests one by one and then plumb that list down through a few branchy/loopy layers of bash scripts.

Having tests committed to git that accidentally don't actually get run in CI is one of my nightmares.  It's difficult to notice when it happens, and can lead to us tagging broken code as stable.

The lists here are a tiny bit brittle in that perhaps someone would add a new test program down the road, but fail to plumb it through correctly.  Hopefully local testing and code review would be enough to notice.

More importantly, though, there is just a lot of boilerplate to document, understand, and keep up to date.  It took me a while to follow the argument-passing conventions and control flow though the code, especially the bash code.

Here's a counter proposal:

(1) The drivers (common, linux, macos) do not hard-code a list of tests.

(2) Instead, all test programs are always run.  The infrastructure code will glob (and sort) files in the `tests` folder and run all of those.

(3) Test that are not relevant (e.g., exercising `apt` commands while on `macOS`) detect as much and embed a trivially-pass short-circuit within the test itself.

In short, we _encapsulate_ the "does this test case apply?" logic within the test itself, instead of the top-level driver.
</blockquote></details>

Done... though I note this isn't different from tests in Bazel. Anyway, it's set up so that `tests/macos` and `tests/linux` can exist with platform-specific tests. (Tests could *also* do like (3), but that forces a somewhat spurious test execution that might be confusing.)

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: feature.

> Reviewed 1 of 1 files at r3, 5 of 6 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056#-:-NEaSkQa1XnWz4Um3MLA:b-n37teh)* status: 6 unresolved discussions, needs at least two assigned reviewers, missing label for release notes (waiting on @mwoehlke-kitware)

___
*[`tools/wheel/test/test-wheel.sh` line 15 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRB0n9rX35KKDe-bT:-NEaRB0oEGkHf9YhHJSO:bbb8n4w) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/test/test-wheel.sh#L15)):*
> ```Shell
> 
> if [[ -z "$1" ]]; then
>     echo "Usage: $0 <test> [<args>]" >&2
> ```

This isn't really optional extra args though, is it?  To my reading, this will always be (and must always be) the path to the wheel.



_Suggestion:_
```Shell
<wheel>
```
___
*[`tools/wheel/wheel_builder/common.py` line 55 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRjOU4NPfXi65JWOx:-NEaRjOV3f3D5ShOnZxm:b-k5lh6u) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L55)):*
> ```Python
> 
> 
> def find_tests(*args):
> ```

This API is so generic to the point of being difficult to follow.

This function should take a single required scalar argument (e.g., named `platform` or `subdir` or something else indicative of its semantics in practice) to help guide the reader in understanding the wheel testing architecture.



_Code quote:_
```Python
*args
```
___
*[`tools/wheel/wheel_builder/common.py` line 57 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRfmoAiGD9UOLklsC:-NEaRfmoAiGD9UOLklsD:bb7phvt) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L57)):*
> ```Python
> def find_tests(*args):
>     """
>     Return list of tests in the common directory and any subdirectories given
> ```

typo



_Suggestion:_
```Python
Returns
```
___
*[`tools/wheel/wheel_builder/common.py` line 59 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaS0O3EMy03e9-2ycP:-NEaS0O3EMy03e9-2ycQ:beqobkn) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L59)):*
> ```Python
>     Return list of tests in the common directory and any subdirectories given
>     as additional arguments.
>     """
> ```

The most important requirement on this function is that it fails-fast when the programmer makes a mistake and accidentally disables all tests.  Therefore:

If the common test directory contains zero tests, then this function must fail-fast.

If the bonus test directory does not exist, then this function must fail-fast.  For now, the platform-specific folders can just have a `README` (so that they exist) in anticipation of adding actual test cases in future PRs.  That probably means only running files name `*.py` instead of all `*`, but that's fine.
___
*[`tools/wheel/wheel_builder/macos.py` line 65 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaQwylASEYNHHf7cPH:-NEaQwylASEYNHHf7cPI:bb7muwv) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/macos.py#L65)):*
> ```Python
>     subprocess.check_call(['bash', install_script, wheel], env=env)
> 
>     # Run individual tests
> ```

punc



_Suggestion:_
```Python
# Run individual tests.
```
___
*[`tools/wheel/wheel_builder/macos.py` line 71 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaQYfI3fj-e1oqfmRR:-NEaQYfI3fj-e1oqfmRS:b9qeauf) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/macos.py#L71)):*
> ```Python
>         subprocess.check_call(['bash', test_script, test, wheel], env=env)
>         print(f'-- Executing test {test} - PASSED')
> ```

The prior `test-wheel` script cleaned up after itself (upon success).  Probably this function should now take on that responsibility?

**Review by mwoehlke-kitware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: 4 unresolved discussions, needs at least two assigned reviewers, missing label for release notes (waiting on @jwnimmer-tri)

___
*[`tools/wheel/test/test-wheel.sh` line 15 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRB0n9rX35KKDe-bT:-NEajsB09AVYfKUsf9iK:b-jbi1zm) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/test/test-wheel.sh#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This isn't really optional extra args though, is it?  To my reading, this will always be (and must always be) the path to the wheel.
</blockquote></details>

I'm not sure how to answer that :slightly_smiling_face:. From the perspective of *this script*, it doesn't know whether a test requires arguments or not. That's a function of the individual test(s).
___
*[`tools/wheel/wheel_builder/common.py` line 57 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRfmoAiGD9UOLklsC:-NEami6q7IOh5TCI8fK6:b-u6yx0t) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L57)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Done. (Will appear in next push.)
___
*[`tools/wheel/wheel_builder/common.py` line 59 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaS0O3EMy03e9-2ycP:-NEakWkM6U7oxIA8Hnor:b5r76jh) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L59)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The most important requirement on this function is that it fails-fast when the programmer makes a mistake and accidentally disables all tests.  Therefore:

If the common test directory contains zero tests, then this function must fail-fast.

If the bonus test directory does not exist, then this function must fail-fast.  For now, the platform-specific folders can just have a `README` (so that they exist) in anticipation of adding actual test cases in future PRs.  That probably means only running files name `*.py` instead of all `*`, but that's fine.
</blockquote></details>

I am deliberately not assuming tests are Python. A test might be `.sh`. I considered only returning executable files, but that seems to have the same danger as forgetting to list a test. Right now, the intended disambiguator is that if a test needs additional files, they can/should be placed in a subdirectory.

That necessarily breaks the idea of changing the function to only accept one directory for additional tests. We could, for now, simply remove the extra directories from the call sites and re-add them when we have platform-specific tests.
___
*[`tools/wheel/wheel_builder/macos.py` line 65 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaQwylASEYNHHf7cPH:-NEamC8NCEof3f9YcH7H:b-u6yx0t) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/macos.py#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

punc
</blockquote></details>

Done. (Will appear in next push.)
___
*[`tools/wheel/wheel_builder/macos.py` line 71 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaQYfI3fj-e1oqfmRR:-NEamGLa8O2idPhJr36N:b-8r1sm8) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/macos.py#L71)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The prior `test-wheel` script cleaned up after itself (upon success).  Probably this function should now take on that responsibility?
</blockquote></details>

Maybe, although I deliberately dropped that as the *build* doesn't clean up after itself either. WDYT, should we clean up only the test directory? Change the builder to clean up everything? (I'm somewhat more inclined to the latter, though I'd probably want to kick that to a new PR.)

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: 4 unresolved discussions, needs at least two assigned reviewers, missing label for release notes (waiting on @mwoehlke-kitware)

___
*[`tools/wheel/test/test-wheel.sh` line 15 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRB0n9rX35KKDe-bT:-NEanm3279oCdIZfgfix:b-62xbkd) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/test/test-wheel.sh#L15)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

I'm not sure how to answer that :slightly_smiling_face:. From the perspective of *this script*, it doesn't know whether a test requires arguments or not. That's a function of the individual test(s).
</blockquote></details>

The higher-layer infrastructure makes no provision for bespoke, test-specific arguments here.  The argument will only and always ever be the wheel filename.

If we can communicate that to someone reading this script, it will make this whole contraption a lot easier to follow along with.  Pretending that test programs can have customized arguments is misleading to people reading this wrapper.
___
*[`tools/wheel/wheel_builder/common.py` line 59 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaS0O3EMy03e9-2ycP:-NEanRj0-tRkvR-M8PpR:b-y5yl3t) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L59)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

I am deliberately not assuming tests are Python. A test might be `.sh`. I considered only returning executable files, but that seems to have the same danger as forgetting to list a test. Right now, the intended disambiguator is that if a test needs additional files, they can/should be placed in a subdirectory.

That necessarily breaks the idea of changing the function to only accept one directory for additional tests. We could, for now, simply remove the extra directories from the call sites and re-add them when we have platform-specific tests.
</blockquote></details>

In that case maybe the scanning here excludes `README.*` but nothing else?  A deny-list is probably safer than an allow-list.

I'm also fine with withdrawing the "extra folder" scanning feature from this PR, but it seems like we're going to need it anyway soon so it's probably just as easy to keep it here.
___
*[`tools/wheel/wheel_builder/macos.py` line 71 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaQYfI3fj-e1oqfmRR:-NEanB6WDMgfYYBIJg69:b-yt1ayt) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/macos.py#L71)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Maybe, although I deliberately dropped that as the *build* doesn't clean up after itself either. WDYT, should we clean up only the test directory? Change the builder to clean up everything? (I'm somewhat more inclined to the latter, though I'd probably want to kick that to a new PR.)
</blockquote></details>

It's OK to do it in another PR.

I lean towards "always cleanup" but I don't feel strongly in the case of macOS here since it's such a special case (it's basically only ever run in CI, on an Unprovisioned machine).

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: 4 unresolved discussions, needs at least two assigned reviewers, missing label for release notes (waiting on @mwoehlke-kitware)

___
*[`tools/wheel/wheel_builder/common.py` line 59 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaS0O3EMy03e9-2ycP:-NEb5pN20D30b67S1WZ3:bdv1jnu) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L59)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

In that case maybe the scanning here excludes `README.*` but nothing else?  A deny-list is probably safer than an allow-list.

I'm also fine with withdrawing the "extra folder" scanning feature from this PR, but it seems like we're going to need it anyway soon so it's probably just as easy to keep it here.
</blockquote></details>

OK This is mostly fixed now (marking resolved).  I've opened a smaller thread below for an adjacent problem.
___
*[`tools/wheel/wheel_builder/common.py` line 66 at r6](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEb5iy-2wFyBGdXstw3:-NEb5iy02PAVbjSzKnEV:b8km8bw) ([raw file](https://github.com/robotlocomotion/drake/blob/c2f5fee421f0d018a890e8e8d4eb43eef4c6e903/tools/wheel/wheel_builder/common.py#L66)):*
> ```Python
>         test_dir_full = os.path.join(resource_root, 'test', 'tests', test_dir)
>         if not os.path.isdir(test_dir_full):
>             continue
> ```

This `continue` is defeating the sanity-checking assertion below.  If the directory does not exist at all, that needs to be an exception.

**Review by mwoehlke-kitware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: 4 unresolved discussions, needs at least two assigned reviewers, missing label for release notes (waiting on @jwnimmer-tri)

___
*[`tools/wheel/test/test-wheel.sh` line 15 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRB0n9rX35KKDe-bT:-NEb-SGN53sLBpmfbd_A:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/test/test-wheel.sh#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The higher-layer infrastructure makes no provision for bespoke, test-specific arguments here.  The argument will only and always ever be the wheel filename.

If we can communicate that to someone reading this script, it will make this whole contraption a lot easier to follow along with.  Pretending that test programs can have customized arguments is misleading to people reading this wrapper.
</blockquote></details>

Done.
___
*[`tools/wheel/wheel_builder/common.py` line 66 at r6](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEb5iy-2wFyBGdXstw3:-NEb6_hO4YPYDEomSuf3:b-dp8zxu) ([raw file](https://github.com/robotlocomotion/drake/blob/c2f5fee421f0d018a890e8e8d4eb43eef4c6e903/tools/wheel/wheel_builder/common.py#L66)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This `continue` is defeating the sanity-checking assertion below.  If the directory does not exist at all, that needs to be an exception.
</blockquote></details>

Ah, right, that should just go away now... Done.

**Comment by mwoehlke-kitware** [comment]
*[`tools/wheel/wheel_builder/common.py` line 55 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaRjOU4NPfXi65JWOx:-NEb7h9vD7zqRNz-TJ6X:b-k0ixen) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/common.py#L55)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This API is so generic to the point of being difficult to follow.

This function should take a single required scalar argument (e.g., named `platform` or `subdir` or something else indicative of its semantics in practice) to help guide the reader in understanding the wheel testing architecture.
</blockquote></details>

`args` → `test_subdirs`.

**Review by jwnimmer-tri — COMMENTED** [review]
+@SeanCurtis-TRI for platform review per schedule, please.

> Reviewed 1 of 2 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056#-:-NEbI0ikBeSDPm2qOtcx:bbnvyd5)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform), missing label for release notes (waiting on @mwoehlke-kitware and @SeanCurtis-TRI)

**Review by mwoehlke-kitware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: LGTM missing from assignee SeanCurtis-TRI(platform), missing label for release notes (waiting on @jwnimmer-tri and @SeanCurtis-TRI)

___
*[`tools/wheel/wheel_builder/macos.py` line 71 at r5](https://reviewable.io/reviews/robotlocomotion/drake/18056#-NEaQYfI3fj-e1oqfmRR:-NEbLcsc32wsXSeOT-Aw:bxeqi7a) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a568ed02a9b1cb3b7b7f8164bebfafa417acf5/tools/wheel/wheel_builder/macos.py#L71)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It's OK to do it in another PR.

I lean towards "always cleanup" but I don't feel strongly in the case of macOS here since it's such a special case (it's basically only ever run in CI, on an Unprovisioned machine). 
</blockquote></details>

Ah, on second look, the *builder* does already clean up builds. So it makes sense to go ahead and just relocate this.

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056)* status: LGTM missing from assignee SeanCurtis-TRI(platform), missing label for release notes (waiting on @SeanCurtis-TRI)

**Review by mwoehlke-kitware — COMMENTED** [review]
+(status: do not merge)

I have a tiny tweak to the docs I want to include, but don't want to re-run CI (not the macOS wheel build, at least) for it.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056#-:-NEbNdQ8208IOabo-bOE:b-6zv7kk)* status: LGTM missing from assignee SeanCurtis-TRI(platform), labeled "do not merge", missing label for release notes (waiting on @SeanCurtis-TRI)

**Review by mwoehlke-kitware — COMMENTED** [review]
-(status: do not merge)

I wasn't going to re-run this for just a doc change, but seeing as it's taking longer than I want to stick around tonight...

@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056#-:-NEbV_xmAHc4XYVXq-d2:bpdnwje)* status: LGTM missing from assignee SeanCurtis-TRI(platform), labeled "do not merge" (waiting on @jwnimmer-tri and @SeanCurtis-TRI)

**Review by SeanCurtis-TRI — COMMENTED** [review]
:LGTM:

> Reviewed 3 of 8 files at r1, 1 of 1 files at r3, 2 of 6 files at r4, 1 of 3 files at r6, 1 of 2 files at r7, 2 of 2 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18056#-:-NEben1g8HUMw2IlfZ2M:bmxilld)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),SeanCurtis-TRI(platform) (waiting on @mwoehlke-kitware)

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot mac-big-sur-unprovisioned-clang-wheel-experimental-snopt-mosek-release please
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-mosek-release please
